win32ss/user/winsrv/consrv/frontends/gui/text.c

```diff
@@ -501,8 +501,136 @@ GuiPaintTextModeBufferCJK(PTEXTMODE_SCREEN_BUFFER Buffer,
                           PRECT rcView,
                           PRECT rcFramebuffer)
 {
-    /* FIXME */
-    GuiPaintTextModeBuffer(Buffer, GuiData, rcView, rcFramebuffer);
+    PCONSRV_CONSOLE Console = Buffer->Header.Console;
+    ULONG TopLine, BottomLine, LeftColumn, RightColumn;
+    WORD Attribute;
+    ULONG CursorX, CursorY, CursorHeight;
+    ULONG Line, Column;
+    HBRUSH CursorBrush, OldBrush;
+    HFONT OldFont, NewFont;
+    BOOLEAN IsUnderline;
+    PCHAR_INFO Ptr;
+
+    ConioInitLongRect(rcFramebuffer, 0, 0, 0, 0);
+
+    if (Buffer->Buffer == NULL)
+        return;
+
+    if (!ConDrvValidateConsoleUnsafe((PCONSOLE)Console, CONSOLE_RUNNING, TRUE))
+        return;
+
+    ConioInitLongRect(rcFramebuffer,
+                      Buffer->ViewOrigin.Y * GuiData->CharHeight + rcView->top,
+                      Buffer->ViewOrigin.X * GuiData->CharWidth  + rcView->left,
+                      Buffer->ViewOrigin.Y * GuiData->CharHeight + rcView->bottom,
+                      Buffer->ViewOrigin.X * GuiData->CharWidth  + rcView->right);
+
+    LeftColumn  = rcFramebuffer->left  / GuiData->CharWidth;
+    RightColumn = rcFramebuffer->right / GuiData->CharWidth;
+    if (RightColumn >= (ULONG)Buffer->ScreenBufferSize.X)
+        RightColumn = Buffer->ScreenBufferSize.X - 1;
+
+    TopLine    = rcFramebuffer->top    / GuiData->CharHeight;
+    BottomLine = rcFramebuffer->bottom / GuiData->CharHeight;
+    if (BottomLine >= (ULONG)Buffer->ScreenBufferSize.Y)
+        BottomLine = Buffer->ScreenBufferSize.Y - 1;
+
+    Attribute = ConioCoordToPointer(Buffer, LeftColumn, TopLine)->Attributes;
+    SetTextColor(GuiData->hMemDC, PaletteRGBFromAttrib(Console, TextAttribFromAttrib(Attribute)));
+    SetBkColor(GuiData->hMemDC, PaletteRGBFromAttrib(Console, BkgdAttribFromAttrib(Attribute)));
+    IsUnderline = !!(Attribute & COMMON_LVB_UNDERSCORE);
+
+    /* Select the new font */
+    NewFont = GuiData->Font[IsUnderline ? FONT_BOLD : FONT_NORMAL];
+    OldFont = SelectObject(GuiData->hMemDC, NewFont);
+
+    for (Line = TopLine; Line <= BottomLine; Line++)
```

Comment:
My suggestions for this function `GuiPaintTextModeBufferCJK()` as well as for `GuiPaintTextModeBuffer()`:
Almost everything can be shared by both functions, so let's call the shared code `GuiPaintTextModeBuffer()` (we reuse the old name). Then only these nested for-loops differ between the old two functions, so the two different implementations of these for-loops can be placed in two helper functions, one for the regular buffer, the other for the CJK . Then these helper functions are called depending on whether the console is in CJK mode.

(see below for the rest of the suggestions...)

Reply:
CJK is slower than non-CJK. I cannot agree with you.

Reply:
I said the for-loops (that are different and make the CJK slower) can go into a separate function, and that separate function gets called from within the main GuiPaintTextModeBuffer when `Console->bIsCJK` check succeeds.

Reply:
Function calling takes the cost of overhead. It makes the processing slower...

Reply:
Commonized.